starters/nextjs-starter-approuter-ts/lib/utils.ts

```diff
@@ -80,3 +80,118 @@ export function getSeoMetadata(
+        contains = true;
+        // Break out of the loop
+        break;
+      }
+    } else {
+      // If the child is an article, check if it matches the article id
+      if (childOfChild.id === article.id) {
+        contains = true;
+        // If it does, append the result's category tree to the current category tree
+        categoryTree.push(child.name);
+        // Break out of the loop
+        break;
+      }
+    }
+  }
+
+  return { contains, categoryTree };
+}
+
+export function getArticlePathFromContentStrucuture(
+  article: Partial<Article> & Pick<Article, "id">,
+  site: Site,
+) {
+  const defaultPath: string[] = [];
+  // If the site is not defined or the content structure is not defined or if the active key is not defined, return the default path
+  if (!site || !site.contentStructure || !site.contentStructure.active) {
+    return defaultPath;
+  }
+  // If the active key is present, it will be an array of objects. Its structure is as follows:
+  // {
+  //   "id": "string",
+  //   "name": "string",
+  //   "type": "string"
+  //   "children": [
+  //     {
+  //       "id": "string",
+  //       "name": "string",
+  //       "type": "string"
+  //     }
+  //   ]
+  // }
+  // type will be one of the following: "category" or "article"
+  // We need to find the article object that contains the articleId
+  const active = site.contentStructure.active;
+  if (typeof active !== "object" || !Array.isArray(active) || !active.length) {
+    return defaultPath;
+  }
+  // Iterate over the active array
+  for (const category of active) {
+    // The categories can be nested, so we need to find the relevant list of categories that contain the articleId
+    // We need to iterate over all the categories, do the same for all its children. Keep doing this until we find the articleId
+    const { contains, categoryTree } = doesCHildContainArticle(
+      category,
+      article,
+    );
+    if (!contains) {
+      continue;
+    }
+    // If the item is found, return the path as a normalized path
+    if (categoryTree && categoryTree.length > 0) {
+      // normalise the name of each category in the categoryTree
+      const normalisedCategoryTree = categoryTree.map((category) =>
+        category
+          .replace(/ /g, "-")
+          .replace(/[^a-zA-Z0-9-]/g, "")
+          .toLowerCase(),
+      );
+      // Return it
+      return normalisedCategoryTree;
+    }
+  }
+
+  return defaultPath;
+}
+
+export function getArticleURLFromSite(
+  article: Partial<Article> & Pick<Article, "id">,
+  site: Site,
+  basePath = "/articles",
+) {
+  // Get the article path
+  const articlePath = getArticlePathFromContentStrucuture(article, site);
+  // Add the basePath before the articlePath and the article slug or id at the end
+  if (articlePath.length > 0) {
+    return `${basePath}/${articlePath.join("/")}/${article.slug || article.id}`;
+  }
+  return `${basePath}/${article.slug || article.id}`;
+}
```

Comment:
I think this should be exposed as a default behavior, so a developer could write
`resolvePath: generateFullArticlePath`

With the idea that there would be other default behaviors like
`generateShallowArticlePath`

Maybe it would be easier to have a single function which has a max depth (0 would be only the article's slug, 1 would be slug + immediate parent, etc.).

The other things I want to achieve with these changes:
- Keep the surface area for implementing developers as simple to use
- We handle the maintenance & testing of these functions
- Developers automatically get improvements by SDK updates